### PR TITLE
Ground Station V2

### PR DESCRIPTION
**Pull request by di-math**
Hey Guys,

this would be the release of the SW-version that I tested on Thursday/Friday, and which turned out to be working pretty well.
In general, I implemented the requested heartbeat and filter feature. Also I added the fix for the UNIX-socket, where the messages where not separated correctly. Lastly I also added that the webserver polls for the UNIX-socket, if it does not exist yet, which makes it easier to startup the bridge and webserver together on the PI, without the webserver just crashing because the UNIX-socket may not exist yet.

If anyone wants to have a look at it...

### CODE REVIEW

File: test/test_unix_socket/receiver.py [hunk 1]
```diff
@@ -23,23 +23,18 @@ def run_dummy_listener(socket_path=DEFAULT_UNIX_SOCKET_PATH):
...
-                    print("No data received. Closing connection.")
-                    client_socket.close()
-                    exit()
-
-                # Decode the received data from bytes to JSON
-                json_data = json.loads(received_data.decode())
+            # Receive data from the server
+            buffer = buffer + client_socket.recv(1024).decode()
+            blocks = buffer.split(chr(23))
+            buffer = blocks.pop()
```
Comment (J0las) [inline test/test_unix_socket/receiver.py:31]: Why not check if the resulting byte object is empty to verify that the connection is still open?

File: test/test_unix_socket/receiver.py [hunk 2]
```diff
@@ -23,23 +23,18 @@ def run_dummy_listener(socket_path=DEFAULT_UNIX_SOCKET_PATH):
...
+
+            # Decode the received data from bytes to JSON
+            for elem in list(blocks):
+                json_data = json.loads(elem)
                 print("Received JSON data:", json_data)
+                blocks.remove(elem)
```
Comment (J0las) [inline test/test_unix_socket/receiver.py:37]: Iterating through the list and removing items from the list inside the loop is sketchy at the very least. This should invalidate the iterator.

File: test/test_unix_socket/receiver.py [hunk 3]
```diff
@@ -23,23 +23,18 @@ def run_dummy_listener(socket_path=DEFAULT_UNIX_SOCKET_PATH):
         client_socket.connect(socket_path)
         print("Connected to server.")
 
+        buffer = ""
         while True:
```
Comment (J0las) [inline test/test_unix_socket/receiver.py:27]: This looks CPU intensive. Does recv block until data is available or does it just return nothing?

File: test/test_unix_socket/receiver.py [hunk 4]
```diff
@@ -23,23 +23,18 @@ def run_dummy_listener(socket_path=DEFAULT_UNIX_SOCKET_PATH):
...
-                # If no data is received, close the connection
-                if not received_data:
-                    print("No data received. Closing connection.")
-                    client_socket.close()
-                    exit()
-
-                # Decode the received data from bytes to JSON
-                json_data = json.loads(received_data.decode())
+            # Receive data from the server
+            buffer = buffer + client_socket.recv(1024).decode()
```
Comment (J0las) [inline test/test_unix_socket/receiver.py:29]: I would explicitly specify the encoding

File: test/test_unix_socket/receiver.py [hunk 5]
```diff
@@ -23,23 +23,18 @@ def run_dummy_listener(socket_path=DEFAULT_UNIX_SOCKET_PATH):
...
+            blocks = buffer.split(chr(23))
+            buffer = blocks.pop()
+
+            # Decode the received data from bytes to JSON
+            for elem in list(blocks):
```
Comment (J0las) [inline test/test_unix_socket/receiver.py:34]: Why not use async with yield to separate the receiving and sending into their own "threads"

File: test/test_unix_socket/receiver.py [hunk 6]
```diff
@@ -23,23 +23,18 @@ def run_dummy_listener(socket_path=DEFAULT_UNIX_SOCKET_PATH):
...
+                json_data = json.loads(elem)
                 print("Received JSON data:", json_data)
+                blocks.remove(elem)
 
     except Exception as e:
```
Comment (J0las) [inline test/test_unix_socket/receiver.py:39]: Don't we want finer grained exception handling.
If there are any issues currently the telemetry node will end.
I would prefer that code tries to stay functional as long as possible and ignore encoding errors in at least a single message.

File: test/test_unix_socket/receiver.py
Comment (J0las) [inline test/test_unix_socket/receiver.py:1]: At least the `import select` is not needed anymore

File: test/test_unix_socket/sender.py
Comment (J0las) [inline test/test_unix_socket/sender.py:1]: Either the argv check or the index uses the wrong value

File: test/test_unix_socket/sender.py [hunk 7]
```diff
@@ -69,8 +69,31 @@ def run_dummy_server(socket_path=DEFAULT_UNIX_SOCKET_PATH):
...
+                            i += 1
+                            time.sleep(2)
+                            client_socket.sendall(
+                                json.dumps(tm.heartbeat_node2(i, True)).encode() + b"\x17")
+                            print("Send message.")
+                            i += 1
+                            time.sleep(3)
 
                     except Exception as e:
                         print("Error sendind data:", e)
```
Comment (J0las) [inline test/test_unix_socket/sender.py:99]: If this is a test why not rethrow the exception

File: web/backend/app.js [hunk 8]
```diff
@@ -20,29 +23,56 @@ if (process.env.HTTPS_CERT_PATH && process.env.HTTPS_KEY_PATH) {
...
+const startServer = async () => {
+  const UNIX_SOCKET_PATH = process.env.UNIX_SOCKET_PATH || "/tmp/thi_drone";
+  const CHECK_INTERVAL = 1000;
+
+  await waitForSocket(UNIX_SOCKET_PATH, CHECK_INTERVAL);
+
+  console.log(`Socket ${UNIX_SOCKET_PATH} exists. Proceeding to connect...`);
+
+  // Connect to the UNIX-socket
+  const UNIX_client_socket = net.createConnection(UNIX_SOCKET_PATH);
```
Comment (J0las) [inline web/backend/app.js]: What happens if the socket gets closed right after the check?

File: web/backend/app.js [hunk 9]
```diff
@@ -20,29 +23,56 @@ if (process.env.HTTPS_CERT_PATH && process.env.HTTPS_KEY_PATH) {
...
+
+  // Start to listen with server
+  const port = process.env.PORT || 3000;
```
Comment (J0las) [inline web/backend/app.js]: Why use port 3000 as the default.
Wouldn't make sense to bind to 80 using sudo and then dropping privileges or the official http port alternative 8080.
Or use an ephemeral port from the range 49152–65535
Port 3000 could be in use by some ruby or other stuff.

https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers

File: web/frontend/src/components/HeartbeatIndicator.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+  useEffect(() => {
+    const interval = setInterval(() => setNow(Date.now() / 1000), 1000);
+    return () => {
+      clearInterval(interval);
+    };
+  }, []);
+
+  const nodeHeartbeats = Object.values(receivedHeartbeats).map((heartbeat, index) => {
+    const nodeOnline = now - heartbeat.timestamp < maxTimeBetweenHeartbeats;
```
Comment (J0las) [inline web/frontend/src/components/HeartbeatIndicator.tsx]: I would not rely on time sync between the time stamp of the node on the pi and any other device.
Maybe wait for the first message and use its timestamp to calculate the offset between the clocks. This will of cause not account for clock drift.
Maybe I am reading this code wrong but I will not learn JS for this.

File: web/frontend/.env.dev [hunk 11]
```diff
@@ -1 +1 @@
-VITE_BACKEND=
\ No newline at end of file
+VITE_BACKEND=192.168.0.101:80  # Standart static IP of the Ground Station Raspi
```
Comment (J0las) [inline web/frontend/.env.dev]:
```suggestion
VITE_BACKEND=192.168.0.101:80  # Standard static IP of the Ground Station Raspi
```

File: web/backend/app.js
Comment (J0las) [inline web/backend/app.js:1]: What about running multiple instances for multiple UAVs in different namespaces. Is the support already present & how can the end user differentiate between the sessions in the browser?

**Review by J0las — COMMENTED** [review]
This is my feedback. I can't really say anything about the java script and web stuff. I prefer _real_ programming languages.

**Comment by J0las** [comment]
TLDR. Please redo in assembly. We can't afford the new react compiler.

File: test/test_unix_socket/receiver.py (hunk 1 above)
Comment (di-math) [reply test/test_unix_socket/receiver.py:31]: recv should block: https://man7.org/linux/man-pages/man2/recv.2.html

Also this is just a testing script for seeing if stuff is sliding, if you know what I mean.

File: test/test_unix_socket/receiver.py (hunk 2 above)
Comment (di-math) [reply test/test_unix_socket/receiver.py:37]: See l. 34 - I am operating on a new list.

File: test/test_unix_socket/receiver.py (hunk 3 above)
Comment (di-math) [reply test/test_unix_socket/receiver.py:27]: See above. tldr: yes, it blocks

File: test/test_unix_socket/receiver.py (hunk 4 above)
Comment (di-math) [reply test/test_unix_socket/receiver.py:29]: Well, yeah. Default is utf-8. I will watch out if somebody might start using Windows-1256 ;)

File: test/test_unix_socket/receiver.py (hunk 5 above)
Comment (di-math) [reply test/test_unix_socket/receiver.py:34]: Just a testing script, for seeing if the UNIX-socket does his (her?) job.

File: web/backend/app.js (hunk 9 above)
Comment (di-math) [reply web/backend/app.js]: This will run inside a container and be remapped.

File: web/frontend/src/components/HeartbeatIndicator.tsx (hunk 10 above)
Comment (di-math) [reply web/frontend/src/components/HeartbeatIndicator.tsx]: No, you're actually right. This is the bug I looked for. I will fix this by saving my own system time on receive of the message. 
This is TS and not JS btw. ;)

File: web/frontend/.env.dev (hunk 11 above)
Comment (di-math) [reply web/frontend/.env.dev]: Thanks man, appreciate it.

File: web/backend/app.js
Comment (di-math) [reply web/backend/app.js:1]: It scales if your purse scales. Just buy new gear.

**Comment by di-math** [comment]
> TLDR. Please redo in assembly. We can't afford the new react compiler.

I will personally write into registers using my bare hands.

File: web/backend/app.js (hunk 8 above)
Comment (di-math) [reply web/backend/app.js]: Yeah you're right. The telemetry_node would restart, and I implemented 29eff01

**Review by Philipp66904 — APPROVED** [review]
As long as it works, I don't really care about the details 😅

**Comment by J0hn-Lawrence** [comment]
Is this the exact version we tested in the lab previously?

**Comment by di-math** [comment]
> Is this the exact version we tested in the lab previously?

Yes, + minor Bug Fix, but I tested these

**Review by J0hn-Lawrence — APPROVED** [review]
Approved, as we tested this in the lab and it works. I recommend to resolve all conversations before merging, so that @J0las is happy as well.

**Review by maf0115 — APPROVED** [review]
If it was tested and it works, it can be marked as approved for me as well.
